actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -89,19 +97,31 @@ class ActionCable.Connection
 
     open: ->
       ActionCable.log("WebSocket onopen event")
-      @disconnected = false
-      @subscriptions.reload()
+      # This is in the case that the browser client opens
+      # the connection regardless of whether a protocol was 
+      # negotiated successfully or not.
+      if not @isProtocolSupported()
+        @close()
+      else
+        ActionCable.log("WebSocket using '" + @protocol() + "' subprotocol.")
+        @disconnected = false
+        @subscriptions.reload()
 
-    close: ->
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+      if not @isProtocolSupported()
+        ActionCable.log("WebSocket is permanently closing because there is no common subprotocol. Client supported protocols: " + protocols.join(','))
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+      else
+        @disconnect(new ActionCable.Events.Disconnected(event.code, event.reason, true))
 
     error: ->
       ActionCable.log("WebSocket onerror event")
-      @disconnect()
```

Comment:
>  I removed `@disconnect()` from the `error` handler for the websocket. From what I understand, this always gets fired with `close` anyways, but it creates a race condition and doesn't provide any information.

This assumes every `error` results in a `close`, yeah? If you can verify then :+1:, but otherwise it makes me a little uncomfortable.